[feature] Automatic provisioning of subnets and IPs

pandafy opened the pull request:
Closes #400

pandafy commented:
### Instructions for testing 

#### Step 0: Pre-requisites

1. Install dependencies 
2. Run migrations

#### Step 1: Create a master subnet
![image](https://user-images.githubusercontent.com/32094433/110534658-57298f00-8145-11eb-9030-38faa821b15f.png)


#### Step 2: Create a Subnet Division Rule
![image](https://user-images.githubusercontent.com/32094433/110535165-ef277880-8145-11eb-9ff6-b064fb3b9593.png)


#### Step 3: Create a VPN Server
Ensure that you have selected following subnet and IP.
![Screenshot_2021-03-10 Change VPN server OpenWISP Admin(1)](https://user-images.githubusercontent.com/32094433/110535636-8987bc00-8146-11eb-9966-bf2b76140b47.png)


#### Step 4: Create a VPN Client Template
![Screenshot_2021-03-10 Add template OpenWISP Admin](https://user-images.githubusercontent.com/32094433/110535865-ce135780-8146-11eb-850e-454b65e687da.png)
![image](https://user-images.githubusercontent.com/32094433/110535946-e08d9100-8146-11eb-9535-0c754ad176fe.png)

#### Step 5: Apply template to a device 
#### Step 6: Verify "System Defined Variables" of the device
![image](https://user-images.githubusercontent.com/32094433/110536432-77f2e400-8147-11eb-8052-193ef05f8a69.png)

pandafy commented:
@nemesisdesign here are my thoughts on how to manage change in  `SubnetDivisionRule` object. I have organized them base on the fields.

### type
We should not allow changes to the type of `SubnetDivisionRule` since that decides when Subnets and IPs are provisioned.

### organization

The organization of the `SubnetDivisionRule` must be the same as that of `master_subnet`. I can't think of a strong reason to prevent users from changing the organization of the `SubnetDivisionRule` but it will require updating `master_subnet` too. We will need to add proper validation for checking this.

### master_subnet

Changing the `master_subnet` will have the following effects

1. All existing Subnets and IPs provisioned using this `SubnetDivisionRule` will be deleted
2. New Subnets and IPs will be provisioned for existing `Config` entries in `SubnetDivisionIndex`.

### size

Changing the `size` of `SubnetDivisionRule` will affect has following effects

1. `{label}_prefixlen` for the rule needs to be updated
2. All existing Subnets and IPs provisioned using this `SubnetDivisionRule` will be deleted
3. New Subnets and IPs will be provisioned for existing `Config` entries in `SubnetDivisionIndex`.

### number_of_subnets

1. All existing Subnets and IPs provisioned using this `SubnetDivisionRule` will be deleted
2. New Subnets and IPs will be provisioned for existing `Config` entries in `SubnetDivisionIndex`.

### number_of_ips

1. All existing IPs provisioned using this `SubnetDivisionRule` will be deleted
2. New IPs will be provisioned for existing `Config` entries in `SubnetDivisionIndex`.

## Implementation Notes

From an implementation standpoint, I think deleting all existing subnets or IPs and then re-creating them will be simple enough for the current implementation. This will not make us deal will free patches of subnets/IPs in the middle.

nemesisdesign reviewed (commented):
Great progress @pandafy :+1: 

> ### type
> We should not allow changes to the type of `SubnetDivisionRule` since that decides when Subnets and IPs are provisioned.

Sounds good, readonly if the object exists.
 
> ### organization

> The organization of the `SubnetDivisionRule` must be the same as that of `master_subnet`. I can't think of a strong reason to prevent users from changing the organization of the `SubnetDivisionRule` but it will require updating `master_subnet` too. We will need to add proper validation for checking this.

I would simply check that the organization is the same or, in case the master subnet is shared, I would allow the master subnet to be shared and the subnet division rule to be defined for a specific org, so potentially multiple orgs can share one subnet if needed, we don't have to make this work right now but we can leave it open, making sure the basics work and deal with it once we need it.

> ### master_subnet
> Changing the `master_subnet` will have the following effects
> 
> 1. All existing Subnets and IPs provisioned using this `SubnetDivisionRule` will be deleted
> 2. New Subnets and IPs will be provisioned for existing `Config` entries in `SubnetDivisionIndex`.

The problem is how do we know if the subnet changed? The post save doesn't tell us which fields have changed.

> ### size
> Changing the `size` of `SubnetDivisionRule` will affect has following effects
> 
> 1. `{label}_prefixlen` for the rule needs to be updated
> 2. All existing Subnets and IPs provisioned using this `SubnetDivisionRule` will be deleted
> 3. New Subnets and IPs will be provisioned for existing `Config` entries in `SubnetDivisionIndex`.

I would allow increasing the size without doing any change but I would not allow shrinking it if there are already subnets created that would fall outside of the new size.

Destroying and recreating the provisioned subnets and IPs sounds like something that we would want to have, but that should be triggered explicitly by users (they may not be happy we do this implicitly, maybe they'll be relying on subnets and IP addresses to stay the same).

> ### number_of_subnets
> 1. All existing Subnets and IPs provisioned using this `SubnetDivisionRule` will be deleted
> 2. New Subnets and IPs will be provisioned for existing `Config` entries in `SubnetDivisionIndex`.

Same here, I would allow to increase it but not reduce it, and if increased I would trigger the provisioning.

> ### number_of_ips
> 1. All existing IPs provisioned using this `SubnetDivisionRule` will be deleted
> 2. New IPs will be provisioned for existing `Config` entries in `SubnetDivisionIndex`.

Same concept as the previous point.

> ## Implementation Notes
> From an implementation standpoint, I think deleting all existing subnets or IPs and then re-creating them will be simple enough for the current implementation. This will not make us deal will free patches of subnets/IPs in the middle.

Would be great to have this as a method, which can be triggered via an admin action and an admin button, but lower priority and only if you want/have time, otherwise we can create an issue for this specific thing.

pandafy commented:
### organization

In the current implementation, I have checked that the master subnet belongs to the organization of the SubnetDivisionRule. I will update this to allow sharing of master subnets across subnet division rules of other organization. :+1:

--------------------

### master_subnet

> The problem is how do we know if the subnet changed? The post save doesn't tell us which fields have changed.

To know that any field of the `SubnetDivisionRule` is changed, we can add a receiver to `pre_save` signal of `SubnetDivisionRule`. In the receiver we can fetch the object from the database using primary key of the object and compare which fields have been changed. Since the changes we need to make are in other models - Subnet, IpAddress and SubnetDivisionIndex - I don't think we will face any problem with that. This is just a thought, I will check whether it works and will report if it doesn't. 

-----------------

### size, number of subnets, number of IPs

> I would allow increasing the size without doing any change but I would not allow shrinking it if there are already subnets created that would fall outside of the new size.

Consider the following example:

```
Master subnet: 10.0.0.0/16
Size: 28
Number of subnets: 2
Number of IPs: 14
```
For above configuration, the first two subnets and IPs will be following: 
 - `10.0.0.0/28` (subnet1)
     - Usable IP range: 10.0.0.1  to 10.0.0.14
 - `10.0.0.16/28` (subnet2)
     - Usable IP range: 10.0.0.17 to 10.0.0.30     

*Now, if the **size is increased to 27** then subnets and IPs will be as following*:
 - `10.0.0.0/27` (subnet1)
     - Usable IP range: 10.0.0.1  to 10.0.0.30
 - `10.0.0.32/27` (subnet2)
     - Usable IP range: 10.0.0.33 to 10.0.0.62

As you can see that some IPs which were earlier generated for subnet2 will come under subnet1 after the change. I think it will be easier(surely not efficient) to delete existing subnets and IPs and re-provision them than to update different relations not to forget variable names. 

**The same applies for increasing number of subnets**

I believe that only increasing/decreasing the number of IPs does not requires huge changes to the database. 

----------------------------

### Implementation Notes

> Would be great to have this as a method, which can be triggered via an admin action and an admin button, but lower priority and only if you want/have time, otherwise we can create an issue for this specific thing.

I think in this PR, we should only allow changing number of IP address and label. For allowing users to change other fields, we will need to make sure that we cover all possible scenarios and add an action to trigger provisioning.

pandafy commented on openwisp_controller/config/base/config.py:
```diff
@@ -428,6 +428,32 @@ def get_vpn_context(self):
 
+    def get_subnet_division_context(self):
+        # NOTE: Use regex to know which subnet division variables
+        # are used in this config and only provide those contexts
```
This is not required now. I overlooked that variables are evaluated in netjsonconfig and here we can't know which variables are used by this configuration.

pandafy commented on openwisp_controller/config/base/vpn.py:
```diff
@@ -141,6 +172,16 @@ def get_context(self):
+                }
+            )
+        if self.ip:
+            c['ip_address'] = self.ip.ip_address
```
You are correct. For now, we can just add the `subnet` field and add `ip` field later.

pandafy commented on openwisp_controller/subnet_division/templates/admin/input_filter.html:
```diff
@@ -0,0 +1,23 @@
+<!-- templates/admin/input_filter.html -->
```
Please ignore this. I copied this code from the filter example shared in the issue description. I will remove it. :+1:

pandafy commented on openwisp_controller/subnet_division/filters.py:
```diff
@@ -0,0 +1,54 @@
+SubnetDivisionIndex = load_model('subnet_division', 'SubnetDivisionIndex')
+
+
+class InputFilter(admin.SimpleListFilter):
```
Added automated tests :raised_hands:

pandafy commented on openwisp_controller/subnet_division/filters.py:
```diff
@@ -0,0 +1,54 @@
+SubnetDivisionIndex = load_model('subnet_division', 'SubnetDivisionIndex')
+
+
+class InputFilter(admin.SimpleListFilter):
```
Added automated tests

pandafy commented on openwisp_controller/subnet_division/filters.py:
```diff
@@ -0,0 +1,54 @@
+        all_choice['query_parts'] = (
+            (k, v)
+            for k, v in changelist.get_filters_params().items()
+            if k != self.parameter_name
```
Fixed!

pandafy commented:
In the recent commits I have made the following changes: 

- Removed IP field from VPN model 
- Added automated tests for SubnetFilter and DeviceFitler by adding tests for SubnetAdmin and DeviceAdmin
- Fixed unclean code
- Added separate celery tasks for updating `SubnetDivisionIndex`, `Subnet` and `IpAddress` objects 
- Developed mechanism for tracking modified fields and updating related objects. Currently only updating `label` and `number_of_ips` is supported.

nemesisdesign reviewed (changes requested):
I sent minor improvements to verbose names and help texts:
https://github.com/openwisp/openwisp-controller/pull/402/commits/f97899af1979ddc8c865937b86ad68f3b403f9a8

### `decoding to str: need a bytes-like object, int found`

Try to save a device config like:

```json
{
    "general": {
        "timezone": "UTC",
        "description": "{{ WG_subnet1 }}, {{WG_subnet1_ip1}}/{{ WG_prefixlen }}, {{WG_subnet1_ip2}}/{{ WG_prefixlen }}"
    }
}
```

I get the following exception:

```
Traceback (most recent call last):
  File "/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/lib/python3.8/site-packages/django/contrib/admin/options.py", line 614, in wrapper
    return self.admin_site.admin_view(view)(*args, **kwargs)
  File "/lib/python3.8/site-packages/django/utils/decorators.py", line 130, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/lib/python3.8/site-packages/django/views/decorators/cache.py", line 44, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/lib/python3.8/site-packages/django/contrib/admin/sites.py", line 233, in inner
    return view(request, *args, **kwargs)
  File "/openwisp-controller/openwisp_controller/config/admin.py", line 134, in change_view
    return super().change_view(request, object_id, form_url, extra_context)
  File "/lib/python3.8/site-packages/django_reversion-3.0.7-py3.8.egg/reversion/admin.py", line 154, in change_view
    return super().change_view(request, object_id, form_url, extra_context)
  File "/lib/python3.8/site-packages/django/contrib/admin/options.py", line 1656, in change_view
    return self.changeform_view(request, object_id, form_url, extra_context)
  File "/lib/python3.8/site-packages/django/utils/decorators.py", line 43, in _wrapper
    return bound_method(*args, **kwargs)
  File "/lib/python3.8/site-packages/django/utils/decorators.py", line 130, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/lib/python3.8/site-packages/django/contrib/admin/options.py", line 1534, in changeform_view
    return self._changeform_view(request, object_id, form_url, extra_context)
  File "/lib/python3.8/site-packages/django/contrib/admin/options.py", line 1579, in _changeform_view
    if all_valid(formsets) and form_validated:
  File "/lib/python3.8/site-packages/django/forms/formsets.py", line 464, in all_valid
    valid &= formset.is_valid()
  File "/lib/python3.8/site-packages/django/forms/formsets.py", line 308, in is_valid
    self.errors
  File "/lib/python3.8/site-packages/django/forms/formsets.py", line 288, in errors
    self.full_clean()
  File "/lib/python3.8/site-packages/django/forms/formsets.py", line 336, in full_clean
    form_errors = form.errors
  File "/lib/python3.8/site-packages/django/forms/forms.py", line 172, in errors
    self.full_clean()
  File "/lib/python3.8/site-packages/django/forms/forms.py", line 376, in full_clean
    self._post_clean()
  File "/lib/python3.8/site-packages/django/forms/models.py", line 405, in _post_clean
    self.instance.full_clean(exclude=exclude, validate_unique=False)
  File "/lib/python3.8/site-packages/django/db/models/base.py", line 1216, in full_clean
    self.clean()
  File "/openwisp-controller/openwisp_controller/config/base/config.py", line 307, in clean
    super().clean()
  File "/openwisp-controller/openwisp_controller/config/base/base.py", line 75, in clean
    backend = self.backend_instance
  File "/lib/python3.8/site-packages/django/utils/functional.py", line 48, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/openwisp-controller/openwisp_controller/config/base/base.py", line 142, in backend_instance
    return self.get_backend_instance()
  File "/openwisp-controller/openwisp_controller/config/base/base.py", line 166, in get_backend_instance
    backend_instance = backend(**kwargs)
  File "/netjsonconfig/netjsonconfig/backends/base/backend.py", line 45, in __init__
    self.config = self._evaluate_vars(self.config, context)
  File "/netjsonconfig/netjsonconfig/backends/base/backend.py", line 94, in _evaluate_vars
    return evaluate_vars(config, context)
  File "/netjsonconfig/netjsonconfig/utils.py", line 93, in evaluate_vars
    data[key] = evaluate_vars(value, context)
  File "/netjsonconfig/netjsonconfig/utils.py", line 93, in evaluate_vars
    data[key] = evaluate_vars(value, context)
  File "/netjsonconfig/netjsonconfig/utils.py", line 108, in evaluate_vars
    data = re.sub(pattern, context[var], data)
  File "/usr/lib/python3.8/re.py", line 210, in sub
    return _compile(pattern, flags).sub(repl, string, count)
  File "/usr/lib/python3.8/re.py", line 327, in _subx
    template = _compile_repl(template, pattern)
  File "/usr/lib/python3.8/re.py", line 318, in _compile_repl
    return sre_parse.parse_template(repl, pattern)
  File "/usr/lib/python3.8/sre_parse.py", line 972, in parse_template
    s = Tokenizer(source)
  File "/usr/lib/python3.8/sre_parse.py", line 228, in __init__
    string = str(string, 'latin1')

Exception Type: TypeError at /admin/config/device/e9081f8d-67c8-470d-850c-eb9c33bd0314/change/
Exception Value: decoding to str: need a bytes-like object, int found
```

Can you try making the resulting value a string instead of an integer to see if this fixes it?
I wonder if it's a bug in netjsonconfig and if it's worth fixing it.

### Admin inline

Can you add the subnet division rule as an inline to Subnet please?

This way we can avoid having a new admin section for this model, which is not really needed:
we can consider this an optional extension of the Subnet model.

**The next is open for dicussion:**

You know also what would be great? If we could add a tiny JS file which
shows only relevant organizations in the subnet division rule
(eg: if the subnet is not shared, we could automatically assign the org and hide the field, if the subnet is shared we can then show the field and all the orgs available), but this makes sense only if we manage to allow having a shared subnet and multiple rules for different orgs, otherwise we can remove the org field altogether and use the org of the subnet.
Actually, now that I'm thinking about this out loud, I am not sure this feature is really needed.
We could keep things simple and just use the org of the master subnet, which can be shared by the way, so we'd just have to find out what happens in that case
and decide if it's acceptable or not.

pandafy commented:
> PS: how does the admin filter works?

@nemesisdesign in following image you can subnets from both devices:
![image](https://user-images.githubusercontent.com/32094433/111514311-9c6a4400-8777-11eb-834c-db6b329e9dc6.png)

Now after filling in the device name in the device filter input, it will show subnets used by "vm" device:
![image](https://user-images.githubusercontent.com/32094433/111514503-d0456980-8777-11eb-964c-440ce7013598.png)

I implemented it from the [blog shared in the issue description](https://hakibenita.com/how-to-add-a-text-filter-to-django-admin).

pandafy commented:
I have added the logic for using sharable subnets in https://github.com/openwisp/openwisp-controller/pull/402/commits/7ab2bb09243ca3939c8fd9467a78e86ac7f5b2d6. In this commit, I have also added a test where vpn_server, template and subnet are shared systemwide. The organization of the generated subnets will be deduced from the organization of device.

nemesisdesign reviewed (changes requested):
Could you resolve the conflict with master please?

nemesisdesign commented on openwisp_controller/subnet_division/templates/admin/input_filter.html:
```diff
@@ -0,0 +1,20 @@
+{% load i18n %}
+
+<h3>{% blocktrans with filter_title=title %} By {{ filter_title }} {% endblocktrans %}</h3>
```
is the space before `By` on purpose?

pandafy replied:
Just to keep code breezy. I checked through the web debugger tool,  all headings of filters have a leading space character. I think it is good to leave it here. 
![Screenshot from 2021-03-26 23-45-27](https://user-images.githubusercontent.com/32094433/112675594-5dca3d00-8e8d-11eb-8396-b71d558d19a5.png)

nemesisdesign commented on openwisp_controller/subnet_division/filters.py:
```diff
@@ -0,0 +1,78 @@
+
+class SubnetFilter(InputFilter):
+    parameter_name = 'subnet'
+    title = 'Subnet'
```
```suggestion
    title = _('subnet')
```

Keep in mind all strings presented to users must be flagged as translatable.
Use non capitalized for consistency.

nemesisdesign commented on openwisp_controller/subnet_division/filters.py:
```diff
@@ -0,0 +1,78 @@
+    """
+
+    parameter_name = 'device'
+    title = 'Device'
```
same here

nemesisdesign commented on openwisp_controller/subnet_division/filters.py:
```diff
@@ -0,0 +1,78 @@
+
+            return queryset.filter(
+                id__in=SubnetDivisionIndex.objects.filter(
+                    config__device__name=device_text
```
I'd simply call `self.value()` here without storing it in a var to keep it shorter since it would be readable anyway

nemesisdesign commented on openwisp_controller/subnet_division/handlers.py:
```diff
@@ -0,0 +1,195 @@
+        division_rule = instance.vpn.subnet.subnetdivisionrule_set.get(
+            organization_id__in=(organization_id, None),
+            type='openwisp_controller.subnet_division.types.VpnSubnetDivisionRuleType',
+        )
```
we have to find a way to move this VPN specific code to the VPN rule type

nemesisdesign commented on README.rst:
```diff
@@ -593,6 +602,18 @@ For example, if we want to change the verbose name to "Hotspot", we could write:
     OPENWISP_CONTROLLER_DEVICE_VERBOSE_NAME = ('Hotspot', 'Hotspots')
 
+``OPENWISP_SUBNET_HIDE_AUTOMATICALLY_GENERATED_SUBNETS_AND_IPS``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
the prefix `OPENWISP_SUBNET` is not consistent with the other settings, we should use `OPENWISP_CONTROLLER`

nemesisdesign commented on README.rst:
```diff
@@ -184,6 +184,15 @@ endpoint, which provides a paginated list of such locations in GeoJSON format:
+~~~~~~~~~~~~~~~~~~~
+
+This app enables controller to automatically allocate subnets and IP addresses to
+devices based on existing subnet division rules created by user.
```
We can shorten to: `This app allows to automatically provision subnets and IP addresses...`

We should add that these subnets and ip addresses will then be available as system defined variables (link the section on system defined variables in the readme) and that the purpose of this app is to allow users to automatically provision and configure specific subnets and ip addresses to the devices without the need of manual intervention.

pandafy replied:
I think users will not be interested in `system defined varaibles` in this section of readme. It will make more sense to it in  **how-to-configure-automatic-provisioning-of-subnets-and-ips** section.

nemesisdesign replied:
ok, but I still suggest to quickly mention that these subnets and IP addresses will be available as configuration variables that can be used in templates, because that provides a better overview of what the features provided and users should be able to understand just by reading this statement, then they can drill down.

nemesisdesign commented on README.rst:
```diff
@@ -939,6 +960,78 @@ Once you have performed the 3 steps above, you can test it as follows:
+
+Voila! You can now use these variables in configuration of the device. Refer to `How to use configuration variables <#how-to-use-configuration-variables>`_
+section of this documentation to learn how to use configuration variables.
+
```
can you please show that the created subnets and ip addresses are available as system defined variables in the device?

Eg:

- show a screenshot in which the system defined variables are visible

Then mentions these variables can be used in templates, provided the default fall back values are provided in the "default values" template field (used mainly for validation).

nemesisdesign commented on openwisp_controller/config/base/vpn.py:
```diff
@@ -141,6 +149,14 @@ def get_context(self):
+            c.update(
+                {
+                    'subnet': str(self.subnet.subnet),
+                    'subnet_prefix': str(self.subnet.subnet.prefixlen),
```
maybe we can call it `subnet_prefixlen` for consistency, what do you think?

pandafy replied:
Made this change! 👍🏼

nemesisdesign commented on openwisp_controller/config/base/vpn.py:
```diff
@@ -141,6 +149,14 @@ def get_context(self):
             c.update([('cert', self.cert.certificate), ('key', self.cert.private_key)])
         if self.dh:
             c.update([('dh', self.dh)])
+        # NOTE: Is this required?
```
it can be useful and is also consistent (why not be able to use the subnet variables in the config? makes sense to me)

nemesisdesign commented on openwisp_controller/subnet_division/admin.py:
```diff
@@ -0,0 +1,118 @@
+        url = reverse(f'admin:{app_label}_device_changelist')
+        if obj.master_subnet is None:
+            return mark_safe(
+                f'<a href="{url}?subnet={str(obj.subnet)}">See all devices</a>'
```
do not forget to flag UX strings as translatable

nemesisdesign commented on openwisp_controller/subnet_division/base/models.py:
```diff
@@ -0,0 +1,117 @@
+
+class AbstractSubnetDivisionRule(TimeStampedEditableModel, OrgMixin):
+    type = models.CharField(max_length=200, choices=app_settings.SUBNET_DIVISION_TYPES)
+    # NOTE: Make master_subnet a cached_property since it is utilized often?
```
just ensure that when it is accessed, the queryset calls `select_related('master_subnet')` so we avoid generating too many queries

nemesisdesign commented on openwisp_controller/subnet_division/handlers.py:
```diff
@@ -0,0 +1,195 @@
+)
+def update_subnet_division_rule_related(sender, instance, created, **kwargs):
+    if created:
+        SubnetDivisionIndex.objects.create(
```
any DB operation done in `post_save` must be executed after the transaction is committed to DB to avoid incurring in issues like https://github.com/openwisp/openwisp-notifications/issues/165

pandafy replied:
Yeah, it occurred to me as well after reading one of your recent review on openwisp-notifications. I have made this change.

nemesisdesign commented on requirements.txt:
```diff
@@ -17,3 +19,4 @@ scp>=0.13.0,<0.14.0
 celery>=4.4.3,<4.5.0
 swapper~=1.1.0
 django-cache-memoize~=0.1
+netaddr~=0.8.0
```
is this still used?

pandafy replied:
Yes, it is. `ipaddr` does not provide a quick way to get the next subnet. I wonder if I can just port this functionality and inherit everything else from `ipaddr`. I kept this to work at the last when everything else is ready.

nemesisdesign replied:
it's ok, we use this library in openwisp-monitoring too.

nemesisdesign reviewed (changes requested):
Great progress @pandafy, see my comments below.

nemesisdesign commented on README.rst:
```diff
@@ -939,6 +961,80 @@ Once you have performed the 3 steps above, you can test it as follows:
+  But don't forget to provide the default fall back values in the "default values" template field
+  (used mainly for validation).
+
 Default Templates
```
I think we should move up default templates and required templates before the sections on variables, because these are more basic features which should be explained first, what do you think?

pandafy replied:
I tried to group it along with other tutorials/how-tos. But you are right, I have reordered them.

nemesisdesign commented on openwisp_controller/config/base/config.py:
```diff
@@ -461,6 +461,34 @@ def get_vpn_context(self):
+            else:
+                context[entry['keyword']] = str(entry['ip__ip_address'])
+
+        # TODO: Find a way to avoid this query
```
is it really avoidable?

pandafy replied:
I added this TODO while building a prototype just to remind me to look for possibility to avoid this query. I put some thought to it and I couldn't find a way to avoid it since we are querying `SubnetDivisionIndex` using `config_id` here. 

Since we will not query the prefixlen variable (e.g. `OW_prefixlen`), I think we can remove this code too and add it back in future if needed. 

https://github.com/openwisp/openwisp-controller/blob/8702de6d89f3423ed0c13c2dd260de928f2fb427/openwisp_controller/subnet_division/handlers.py#L68-L71

nemesisdesign commented on openwisp_controller/subnet_division/apps.py:
```diff
@@ -0,0 +1,27 @@
+        from . import settings as app_settings
+        from .handlers import destroy_provisioned_subnets, provision_subnets_ips
+
+        for (rule_path, name) in app_settings.SUBNET_DIVISION_TYPES:
```
I think these rounded parenthesis are not needed

nemesisdesign commented on openwisp_controller/subnet_division/apps.py:
```diff
@@ -0,0 +1,27 @@
+                receiver=destroy_provisioned_subnets,
+                sender=rule_class.destroyer_sender,
+                dispatch_uid=rule_class.destroyer_dispatch_uid,
+            )
```
nice :+1:

nemesisdesign commented on openwisp_controller/subnet_division/handlers.py:
```diff
@@ -0,0 +1,221 @@
+                    'openwisp_controller.subnet_division.types'
+                    '.VpnSubnetDivisionRuleType'
+                ),
+            )
```
are you sure this query will always return only one result?

Can there be only one rule of a specific type and a specific org?

Part of this function is specific for the VPN rule, can you find a way to extract that part and move it to the vpn rule class?

For the rest of the function, can you find another place where to move this logic? For example a class method in one of the models could be good.

pandafy replied:
> are you sure this query will always return only one result?

Yes, it should.

> Can there be only one rule of a specific type and a specific org?

Yes, I added following constraints as well. It also made logical sense. If someone wants that more subnets or IPs should be provisioned, they should update the rule instead of creating two rules of the same type.
https://github.com/openwisp/openwisp-controller/blob/8702de6d89f3423ed0c13c2dd260de928f2fb427/openwisp_controller/subnet_division/base/models.py#L36-L47

nemesisdesign commented on tests/openwisp2/settings.py:
```diff
@@ -215,6 +222,12 @@
+    )
+    SUBNET_DIVISION_SUBNETDIVISIONINDEX_MODEL = (
+        'sample_subnet_division.SubnetDivisionIndex'
+    )
```
The readme section which explains how to extend ow-controller needs to be updated as well.

nemesisdesign reviewed (changes requested):
the build is failing for a minor style issue, see also my other comments below

nemesisdesign commented on README.rst:
```diff
@@ -1227,7 +1325,7 @@ Configure celery (you may use a different broker if you want):
 
 If you decide to use redis (as shown in these examples),
-install the requierd python packages::
+install the required python packages::
```
I think the docs for extending ow-controller should still  be updated right?

nemesisdesign replied:
@pandafy this is still pending, right?

nemesisdesign commented on openwisp_controller/subnet_division/handlers.py:
```diff
@@ -0,0 +1,111 @@
+            if instance_value != db_value:
+                modified_fields[field.name] = db_value
+
+        _subnet_division_rule_update_queue[str(instance.id)] = modified_fields
```
Let's move this logic to the models, following the structure of the existing code of OpenWISP Controller, if you look in the other apps, you will see signal receivers and celery tasks always call methods on the model, this allows to concentrate the core logic in one place which is good, because using signals usually makes the code harder to read.

It also gives another advantage: the method can be overridden by apps which extend openwisp controller and provide customized models.

nemesisdesign commented on openwisp_controller/subnet_division/handlers.py:
```diff
@@ -0,0 +1,111 @@
+                tasks.provision_extra_ips.delay(
+                    rule_id=str(instance.id),
+                    old_number_of_ips=modified_fields['number_of_ips'],
+                )
```
same here, this could be a model method

nemesisdesign commented on openwisp_controller/subnet_division/handlers.py:
```diff
@@ -0,0 +1,111 @@
+
+    if created:
+        return
+    transaction.on_commit(update_subnet_division_index_entries)
```
which you would pass here

nemesisdesign commented on openwisp_controller/subnet_division/handlers.py:
```diff
@@ -0,0 +1,111 @@
+    def delete_subnets():
+        Subnet.objects.filter(
+            organization_id=instance.organization_id, name__startswith=instance.label
+        ).delete()
```
same here, I'd still make a model method for this because it could be overridden with custom logic

nemesisdesign commented on openwisp_controller/subnet_division/rule_types/base.py:
```diff
@@ -0,0 +1,172 @@
+
+    @classmethod
+    def destroyer_receiver(cls, instance, **kwargs):
+        from ..handlers import destroy_provisioned_subnets
```
why this?

Why not just putting the code here? I don't see `destroy_provisioned_subnets` being called anywhere else, I think the logic belongs to this class.

nemesisdesign commented on openwisp_controller/subnet_division/tasks.py:
```diff
@@ -0,0 +1,76 @@
+
+@shared_task
+def update_subnet_division_index(rule_id):
+    division_rule = SubnetDivisionRule.objects.get(id=rule_id)
```
please handle the exceptional case in which the rule_id is not found, it should not happen but in extreme conditions it can happen as it happened with other celery tasks (eg: the worker queue is full and being processed slowly, then the object is deleted by the user and by the time the queue gets to the task the specific ID is not there anymore), see:
https://github.com/openwisp/openwisp-controller/blob/master/openwisp_controller/config/tasks.py#L19-L25
Please do it similarly (log a warning if this happens).

nemesisdesign commented on openwisp_controller/subnet_division/tasks.py:
```diff
@@ -0,0 +1,76 @@
+
+@shared_task
+def update_subnet_name_description(rule_id):
+    division_rule = SubnetDivisionRule.objects.get(id=rule_id)
```
same here

nemesisdesign commented on openwisp_controller/subnet_division/tasks.py:
```diff
@@ -0,0 +1,76 @@
+            )
+
+    IpAddress.objects.bulk_create(generated_ips)
+    SubnetDivisionIndex.objects.bulk_create(generated_indexes)
```
maybe we can wrap these two calls in a transaction to ensure they're either all created or none gets created (so the operation can be repeated easily in corner cases which need manual intervention).

nemesisdesign commented on openwisp_controller/subnet_division/tasks.py:
```diff
@@ -0,0 +1,76 @@
+    generated_ips = []
+    generated_indexes = []
+
+    division_rule = SubnetDivisionRule.objects.get(id=rule_id)
```
same here to handle the not found exception

nemesisdesign reviewed (changes requested):
Looking very good, some more refinements to do and then we're ready to start testing this in the real world.

nemesisdesign commented on openwisp_controller/subnet_division/tasks.py:
```diff
@@ -0,0 +1,116 @@
+        identifiers[0] = division_rule.label
+        index.keyword = '_'.join(identifiers)
+
+    transaction.on_commit(
```
did I ask you to add `transaction.on_commit` here? I think it's not needed here.. what other save operation is performed in this task that we have to wait for?

pandafy replied:
No, you did not. I will revert the changes. 👍🏼

nemesisdesign commented on openwisp_controller/subnet_division/tasks.py:
```diff
@@ -0,0 +1,116 @@
+            subnet_queryset, fields=['name', 'description'], batch_size=20
+        )
+    )
+
```
same here

nemesisdesign commented on openwisp_controller/subnet_division/tasks.py:
```diff
@@ -0,0 +1,116 @@
+        lambda: _create_ipaddress_and_subnetdivision_index_objects(
+            generated_ips, generated_indexes
+        )
+    )
```
and here

pandafy replied:
> maybe we can wrap these two calls in a transaction to ensure they're either all created or none gets created (so the operation can be repeated easily in corner cases which need manual intervention).

- https://github.com/openwisp/openwisp-controller/pull/402#discussion_r606701794

Added this after this comment from you.

pandafy commented on openwisp_controller/subnet_division/migrations/__init__.py:
```diff
@@ -0,0 +1,33 @@
+
+def assign_permissions_to_groups(apps, schema_editor):
+    create_default_permissions(apps, schema_editor)
+    operators_and_admins_can_manage = ['subnetdivisionrule']
+    admin_manage_operations = ['add', 'change', 'delete', 'view']
+    operator_manage_operations = ['view']
```
@nemesisdesign will you please double check the permissions? 

[From openwisp-ipam](https://github.com/openwisp/openwisp-ipam/blob/03b6a28f7a6cdb1d3d2e3436c7593f3181be94e7/openwisp_ipam/migrations/__init__.py#L19), operators only have permissions to view subnets. And I believe it's logical to maintain that for SubnetDivisionRule.

nemesisdesign replied:
Ensure users with the "operator" group can only view.

nemesisdesign reviewed (changes requested):
I found an issue during testing: after creating a device which uses a VPN having a subnet with a subnet division rule specified, I tried deleting the device and I get this error:

```
Traceback (most recent call last):
  File "/lib/python3.8/site-packages/django/db/models/fields/related_descriptors.py", line 173, in __get__
    rel_obj = self.field.get_cached_value(instance)
  File "/lib/python3.8/site-packages/django/db/models/fields/mixins.py", line 15, in get_cached_value
    return instance._state.fields_cache[cache_name]

During handling of the above exception ('config'), another exception occurred:
  File "/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/lib/python3.8/site-packages/django/contrib/admin/options.py", line 614, in wrapper
    return self.admin_site.admin_view(view)(*args, **kwargs)
  File "/lib/python3.8/site-packages/django/utils/decorators.py", line 130, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/lib/python3.8/site-packages/django/views/decorators/cache.py", line 44, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/lib/python3.8/site-packages/django/contrib/admin/sites.py", line 233, in inner
    return view(request, *args, **kwargs)
  File "/lib/python3.8/site-packages/django/utils/decorators.py", line 43, in _wrapper
    return bound_method(*args, **kwargs)
  File "/lib/python3.8/site-packages/django/utils/decorators.py", line 130, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/lib/python3.8/site-packages/django/contrib/admin/options.py", line 1846, in delete_view
    return self._delete_view(request, object_id, extra_context)
  File "/lib/python3.8/site-packages/django/db/transaction.py", line 276, in __exit__
    connection.set_autocommit(True)
  File "/lib/python3.8/site-packages/django/db/backends/base/base.py", line 420, in set_autocommit
    self.run_and_clear_commit_hooks()
  File "/lib/python3.8/site-packages/django/db/backends/base/base.py", line 651, in run_and_clear_commit_hooks
    func()
  File "/openwisp-controller/openwisp_controller/subnet_division/rule_types/base.py", line 58, in <lambda>
    lambda: cls.destroy_provisioned_subnets_ips(instance, **kwargs)
  File "/openwisp-controller/openwisp_controller/subnet_division/rule_types/base.py", line 178, in destroy_provisioned_subnets_ips
    subnet_ids = instance.config.subnetdivisionindex_set.values_list('subnet_id')
  File "/lib/python3.8/site-packages/django/db/models/fields/related_descriptors.py", line 187, in __get__
    rel_obj = self.get_object(instance)
  File "/lib/python3.8/site-packages/django/db/models/fields/related_descriptors.py", line 154, in get_object
    return qs.get(self.field.get_reverse_related_filter(instance))
  File "/lib/python3.8/site-packages/django/db/models/query.py", line 429, in get
    raise self.model.DoesNotExist(

Exception Type: DoesNotExist at /admin/config/device/a1e4ebed-bc22-4aaf-ae59-6d38693c02ee/delete/
Exception Value: Config matching query does not exist.
```

Can you please also double check why the build is failing for django 3.0 and fix it?

pandafy commented on openwisp_controller/subnet_division/rule_types/base.py:
```diff
@@ -54,9 +54,7 @@ def provision_receiver(cls, instance, **kwargs):
-        transaction.on_commit(
-            lambda: cls.destroy_provisioned_subnets_ips(instance, **kwargs)
-        )
+        cls.destroy_provisioned_subnets_ips(instance, **kwargs)
```
Since deleting subnets was being triggered `on_comit`, the `on_delete=model.CASCADE` set on config field of `SubnetDivisionRuleIndex` was deleting the entries of the subnets to be deleted from the table, before our handler could get executed. This was resulting in empty queryset(for fetching subnets to be deleted). 

Removing that handler from `transaction.on_comit` solved the issue. I have added automated tests to ensure that.

nemesisdesign reviewed (approved):
I think this one is ready to be merged. There may be other issues, but we'll merge this in the 1.0.x branch, which will be the future release of OpenWISP after 0.9.x (which we should be releasing in the next few weeks).

Please merge when you feel ready but squash the commits (we can keep some of the most important details in the extended part of the commit message) and recreate this branch (github will delete it automatically) so the screenshots in the README will keep loading.
